gc/src/gc.rs

```diff
@@ -199,13 +214,23 @@ fn collect_garbage(st: &mut GcState) {
     unsafe fn mark(head: &Cell<Option<NonNull<GcBox<dyn Trace>>>>) -> Vec<Unmarked<'_>> {
         // Walk the tree, tracing and marking the nodes
         let mut mark_head = head.get();
-        while let Some(node) = mark_head {
-            if (*node.as_ptr()).header.roots() > 0 {
-                (*node.as_ptr()).trace_inner();
+        MARK_QUEUE.with(|queue| {
```

Comment:
My understanding is that this `head` is a linked list containing all currently known `Gc`'s effectively, and what we're interested in doing here is only marking + calling into the nested `Trace::trace()` for ones that are rooted, and now we're trying to do that while avoiding "drilling down" into any nested `Gc`'s that we encounter during tracing (but we still need to accomplish marking them and calling into their nested `Trace::trace()`'s)

So the basic idea is to modify `Gc`'s `Trace::trace()` implementation to "push itself onto a queue" that then gets processed as a second pass